workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/Readers/ReaderBase.cs

```diff
@@ -0,0 +1,154 @@
+                    return Authority.Authoritative;
+                }
+
+                if (EntityManager.HasComponent<NotAuthoritative<TSpatialComponentData>>(Entity))
+                {
+                    return Authority.NotAuthoritative;
+                }
+
+                throw new AuthorityComponentNotFoundException($"No authority component found for the entity with index {Entity.Index}.");
+            }
+        }
+
+        public abstract TSpatialComponentData Data { get; }
+
+        private readonly List<GameObjectDelegates.AuthorityChanged> authorityChangedDelegates
+            = new List<GameObjectDelegates.AuthorityChanged>();
+
+        public event GameObjectDelegates.AuthorityChanged AuthorityChanged
+        {
+            add => authorityChangedDelegates.Add(value);
+            remove => authorityChangedDelegates.Remove(value);
+        }
+
+        private readonly List<GameObjectDelegates.ComponentUpdated<TComponentUpdate>> componentUpdateDelegates
+            = new List<GameObjectDelegates.ComponentUpdated<TComponentUpdate>>();
+
+        public event GameObjectDelegates.ComponentUpdated<TComponentUpdate> ComponentUpdated
+        {
+            add => componentUpdateDelegates.Add(value);
+            remove => componentUpdateDelegates.Remove(value);
+        }
+
+        /// <summary>
+        /// Helper method to dispatch property updates to callbacks while forwarding exceptions to the log dispatcher.
+        /// </summary>
+        /// <param name="payload">The value for the property update.</param>
+        /// <param name="callbacks">The property update handlers.</param>
+        /// <typeparam name="T">The property update type.</typeparam>
+        protected void DispatchWithErrorHandling<T>(Option<T> payload, IEnumerable<Action<T>> callbacks)
+        {
+            if (!payload.HasValue)
+            {
+                return;
+            }
+
+            foreach (var callback in callbacks)
+            {
+                try
+                {
+                    callback(payload.Value);
+                }
+                catch (Exception e)
+                {
+                    // Log the exception but do not rethrow it, as other delegates should still get called
+                    logDispatcher.HandleLog(LogType.Exception, new LogEvent().WithException(e));
+                }
+            }
+        }
+
+        public void OnAuthorityChange(Authority authority)
+        {
+            foreach (var callback in authorityChangedDelegates)
+            {
+                try
+                {
+                    callback(authority);
+                }
+                catch (Exception e)
+                {
+                    // Log the exception but do not rethrow it, as other delegates should still get called
+                    logDispatcher.HandleLog(LogType.Exception, new LogEvent().WithException(e));
+                }
+            }
+        }
+
+        public void OnComponentUpdate()
```

Comment:
The current plan is to have the corresponding Update object passed directly to each specific Reader implementation, right? So this either needs to have a generic parameter or should be left for the code generated implementers

Reply:
Added the parameter to the method